File: CloudSeed.Native/MultitapDiffuser.cpp

#include <memory>
#include <array>
#include "MultitapDiffuser.h"
#include "Utils.h"
#include "AudioLib/ShaRandom.h"

namespace CloudSeed
{
	MultitapDiffuser::MultitapDiffuser(int bufferSize)
	{
		len = bufferSize;
		buffer = new double[bufferSize];
		output = new double[bufferSize];
		index = 0;
		count = 1;
		length = 1;
		gain = 1.0;
		decay = 0.0;
		crossSeed = 0.0;
		UpdateSeeds();
	}

	MultitapDiffuser::~MultitapDiffuser()
	{
		delete buffer;
		delete output;
	}

	void MultitapDiffuser::SetSeed(int seed)
	{
		this->seed = seed;
		UpdateSeeds();
	}

	void MultitapDiffuser::SetCrossSeed(double crossSeed)
	{
		this->crossSeed = crossSeed;
		UpdateSeeds();
	}

	double* MultitapDiffuser::GetOutput()
	{ 
		return output;
	}

	void MultitapDiffuser::SetTapCount(int tapCount)
	{
		count = tapCount;
		Update();
	}

	void MultitapDiffuser::SetTapLength(int tapLength)
	{
		length = tapLength;
		Update();
	}

	void MultitapDiffuser::SetTapDecay(double tapDecay)
	{
		decay = tapDecay;
		Update();
	}

	void MultitapDiffuser::SetTapGain(double tapGain)
	{
		gain = tapGain;
		Update();
	}

	void MultitapDiffuser::Process(double* input, int sampleCount)
	{
		// prevents race condition when parameters are updated from Gui
		if (isDirty)
		{
			tapGainsTemp = tapGains;
			tapPositionTemp = tapPosition;
			countTemp = count;
			isDirty = false;
		}

		int* const tapPos = &tapPositionTemp[0];
		double* const tapGain = &tapGainsTemp[0];
		const int cnt = countTemp;

		for (int i = 0; i < sampleCount; i++)
		{
			if (index < 0) index += len;
			buffer[index] = input[i];
			output[i] = 0.0;

			for (int j = 0; j < cnt; j++)
			{
				auto idx = (index + tapPos[j]) % len;
				output[i] += buffer[idx] * tapGain[j];
			}

			index--;
		}
	}

	void MultitapDiffuser::ClearBuffers()
	{
		Utils::ZeroBuffer(buffer, len);
		Utils::ZeroBuffer(output, len);
	}
	
	void MultitapDiffuser::Update()
	{
		vector<double> newTapGains;
		vector<int> newTapPosition;

		int s = 0;
		auto rand = [&]() {return seedValues[s++]; };

		if (count < 1)
			count = 1;

		if (length < count)
			length = count;

		// used to adjust the volume of the overall output as it grows when we add more taps
		double tapCountFactor = 1.0 / (1 + std::sqrt(count / MaxTaps));

		newTapGains.resize(count);
		newTapPosition.resize(count);

		vector<double> tapData(count, 0.0);

		auto sumLengths = 0.0;
		for (size_t i = 0; i < count; i++)
		{
			auto val = 0.1 + rand();
			tapData[i] = val;
			sumLengths += val;
		}

		auto scaleLength = length / sumLengths;
		newTapPosition[0] = 0;

		for (int i = 1; i < count; i++)
		{
			newTapPosition[i] = newTapPosition[i - 1] + (int)(tapData[i] * scaleLength);
		}

		double sumGains = 0.0;
		double lastTapPos = newTapPosition[count - 1];
		for (int i = 0; i < count; i++)
		{
			// when decay set to 0, there is no decay, when set to 1, the gain at the last sample is 0.01 = -40dB
			auto g = std::pow(10, -decay * 2 * newTapPosition[i] / (double)(lastTapPos + 1));
			
			auto tap = (2 * rand() - 1) * tapCountFactor;
			newTapGains[i] = tap * g * gain;
		}

		// Set the tap vs. clean mix
		newTapGains[0] = (1 - gain);

		this->tapGains = newTapGains;
		this->tapPosition = newTapPosition;
		isDirty = true;
	}
	
	void MultitapDiffuser::UpdateSeeds()
	{
		this->seedValues = AudioLib::ShaRandom::Generate(seed, 100, crossSeed);
		Update();
	}
}

File: CloudSeed/MultitapDiffuser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AudioLib;

namespace CloudSeed
{
	public class MultitapDiffuser
	{
		private readonly double[] buffer;
		private readonly double[] output;
		private readonly int len;

		private int index;
		private double[] tapGains;
		private int[] tapPosition;
		private double[] seeds;

		private int count;
		private double length;
		private double gain;
		private double decay;

		public MultitapDiffuser(int bufferSize)
		{
			this.len = bufferSize;
			this.buffer = new double[bufferSize];
			this.output = new double[bufferSize];
			index = 0;
			Seeds = new ShaRandom().Generate(1, 100).ToArray();
		}

		public double[] Seeds
		{
			get { return seeds; }
			set
			{
				seeds = value;
				Update();
			}
		}

		public double[] Output { get { return output; } }

		public void SetTapCount(int tapCount)
		{
			count = tapCount;
			Update();
		}

		public void SetTapLength(int tapLength)
		{
			length = tapLength;
			Update();
		}

		public void SetTapDecay(double tapDecay)
		{
			decay = tapDecay;
			Update();
		}

		public void SetTapGain(double tapGain)
		{
			gain = tapGain;
			Update();
		}

		public void Update()
		{
			int s = 0;
			Func<double> rand = () => Seeds[s++];

			if (count < 1)
				count = 1;

			if (length < count)
				length = count;

			tapGains = new double[count];
			tapPosition = new int[count];

			var tapData = Enumerable.Range(0, count).Select(x => 0.1 + rand()).ToArray();
			var scale = length / tapData.Sum();
			tapPosition[0] = 0;

			for (int i = 1; i < count; i++)
			{
				tapPosition[i] = tapPosition[i - 1] + (int)(tapData[i] * scale);
			}

			for (int i = 0; i < count; i++)
			{
				var g = gain * (1 - decay * (i / (double)count));
				tapGains[i] = g * (2 * rand() - 1);
			}

			tapGains[0] = (1 - gain) + tapGains[0] * gain;
		}

		public void Process(double[] input, int sampleCount)
		{
			var taps = tapGains.Length;

			for (int i = 0; i < sampleCount; i++)
			{
				if (index < 0) index += len;
				buffer[index] = input[i];
				output[i] = 0.0;

				for (int j = 0; j < taps; j++)
				{
					var idx = (index + tapPosition[j]) % len;
					output[i] += buffer[idx] * tapGains[j];
				}

				index--;
			}
		}

		public void ClearBuffers()
		{
			buffer.Zero();
		}
	}
}
